### PR TITLE
Prevent infinite loops

### LINKED ISSUE #124: Prevent component code from running infinite loops
Since the components are (currently) executed inside the same process as the registry, a poorly (or maliciously) written component that contains an infinite loop can bring the registry to it's knees (i.e. cause the execution to get stuck inside a never-ending thread).

We could use something like burrito to walk the `server.js` code for a component (either at publish time or execution time) to inject code that can dump out and return control in the event that a loop runs for "a long time".

```javascript
var burrito = require('burrito');

var serverjs = "module.exports.data=function(){ var x=0; while(true){ x = 123; } for(var i = 0; i < 1e10; i++) { console.log('blarg'); } callback(null, {}); };";

var CONST_MAX_ITERATIONS = 1e7;
var monitoredKeywords = ['while', 'for', 'do'];

var src = burrito(serverjs, function (node) {
  if(monitoredKeywords.indexOf(node.name) > -1){
    node.wrap('{ var __ITER = ' + CONST_MAX_ITERATIONS + ';  %s }');
  }

  if(node.parent() && monitoredKeywords.indexOf(node.parent().name) > -1 && node.name === 'block'){
    node.wrap('{ if(__ITER <=0){ throw new Error("loop exceeded maximum allowed iterations"); }  %s __ITER--; }');
  }
});

```

This would wrap the loops in a block and give them an iterator and a check:

```
module.exports.data = function() {
    var x = 0;
    {
        var __ITER = 1e7;
        while (true) {
            if (__ITER <= 0) {
                throw new Error("loop exceeded maximum allowed iterations");
            }
            {
                x = 123;
            }
            __ITER--;
        }
    }
    {
        var __ITER = 1e7;
        for (var i = 0; i < 1e10; i++) {
            if (__ITER <= 0) {
                throw new Error("loop exceeded maximum allowed iterations");
            }
            {
                console.log("blarg");
            }
            __ITER--;
        }
    }
    callback(null, {});
};
```

### PR DESCRIPTION
**Pull request by andyroyle**
closes #124 

This approach doesn't necessarily cope with nested loops, insofar as the `__ITER` variable will get redeclared in the lower scope. 

This means that nested loops will run until the bottom loop hits the limit (not necessarily just the outer). We can guard separately against nested loops if we want to.

### CODE REVIEW

**Comment by matteofigus** [comment]
This LGTM, but given it is a quite critical functionality, I would actually like if @jankowiakmaria and @pbazydlo could have a look too.

File: cli/domain/package-server-script.js [hunk 1]
```diff
@@ -5,20 +5,41 @@ var format = require('stringformat');
...
 var path = require('path');
 var uglifyJs = require('uglify-js');
+var falafel = require('falafel');
 var _ = require('underscore');
 
 var hashBuilder = require('../../utils/hash-builder');
 var strings = require('../../resources');
 
+var wrapLoops = function(code){
+  var CONST_MAX_ITERATIONS = 1e9; // should this be configurable?
```
Comment (matteofigus) [inline cli/domain/package-server-script.js]: Best would probably be to have a default (used during dev) and then get a configurable value on the registry level.

During publishing, CLI would ask the registry for packaging config and then use that value. 
But at the moment, I don't care, this is enough.
Can you just remove the comment and perhaps move the value here? https://github.com/opentable/oc/blob/master/resources/settings.js

File: test/unit/cli-domain-package-server-script.js [hunk 2]
```diff
@@ -36,7 +28,7 @@ var initialise = function(fs, uglifyStub){
         return _.toArray(arguments).join('/');
       }
     }
-  });
+  }, { console: console });
```
Comment (matteofigus) [inline test/unit/cli-domain-package-server-script.js]: can we remove this?

**Comment by andyroyle** [comment]
@matteofigus I've made those changes.

@jankowiakmaria how scary does it look?

**Comment by jankowiakmaria** [comment]
@andyroyle :fearful:

**Comment by andyroyle** [comment]
@matteofigus @jankowiakmaria tell you what, I'll pull the source of all the existing components and run it though the new packager and see if anything shakes loose

**Comment by matteofigus** [comment]
Thanks, that's actually a good idea

**Comment by andyroyle** [comment]
I've run it against a selection of the components in our Preprod registry; Some I couldn't verify because the components don't link to the repo, but so far nothing I've seen worries me.

**Comment by andyroyle** [comment]
@matteofigus so I can't see any immediate problems that will be caused by this.

What do you think? Merge?

**Comment by matteofigus** [comment]
Let me bump the patch for the other PR I just merged, then I would do a merge + minor for this, ok? give me a couple of mins.

**Comment by matteofigus** [comment]
Ok, published in `0.18.0`

**Pull request merged**